### DISCUSSION TITLE
v3.5.9: Being asked to reallocate Port 2380 when creating 2nd node for a multi-node etcd cluster in Docker container

### DISCUSSION THREAD

**Discussion by dev-vaayen** [etcd-io/etcd#16130] [Q&A]
Hi everyone

I am following [this section](https://etcd.io/docs/v3.5/op-guide/container/#running-a-3-node-etcd-cluster-1) of the etcd v3.5 document for creating a 3 node etcd cluster using Docker.

The issue that I am getting is that after creating the first node, I am not able to create the next nodes as I am hit with the error provided below, asking me to change the 2380 port for the second node:
```
docker: Error response from daemon: driver failed programming external connectivity on endpoint etcd-node2 (9043ccd4bc718543823be6dad91e9689ea4d9a5a6198be6488429ee36d8e5a9a): Bind for 0.0.0.0:2380 failed: port is already allocated.
```
Please correct me if I am wrong but aren't the nodes supposed to be on the same ports? If no, then should I change the ports? Otherwise please provide a solution for how to run the next nodes on the same port.

For reference, the code that I followed from the above documentation page was as follows:
```
REGISTRY=quay.io/coreos/etcd
# available from v3.2.5
REGISTRY=gcr.io/etcd-development/etcd

# For each machine
ETCD_VERSION=latest
TOKEN=my-etcd-token
CLUSTER_STATE=new
NAME_1=etcd-node-0
NAME_2=etcd-node-1
NAME_3=etcd-node-2
HOST_1=10.20.30.1
HOST_2=10.20.30.2
HOST_3=10.20.30.3
CLUSTER=${NAME_1}=http://${HOST_1}:2380,${NAME_2}=http://${HOST_2}:2380,${NAME_3}=http://${HOST_3}:2380
DATA_DIR=/var/lib/etcd

# For node 1
THIS_NAME=${NAME_1}
THIS_IP=${HOST_1}
docker run \
  -p 2379:2379 \
  -p 2380:2380 \
  --volume=${DATA_DIR}:/etcd-data \
  --name etcd ${REGISTRY}:${ETCD_VERSION} \
  /usr/local/bin/etcd \
  --data-dir=/etcd-data --name ${THIS_NAME} \
  --initial-advertise-peer-urls http://${THIS_IP}:2380 --listen-peer-urls http://0.0.0.0:2380 \
  --advertise-client-urls http://${THIS_IP}:2379 --listen-client-urls http://0.0.0.0:2379 \
  --initial-cluster ${CLUSTER} \
  --initial-cluster-state ${CLUSTER_STATE} --initial-cluster-token ${TOKEN}

# For node 2
THIS_NAME=${NAME_2}
THIS_IP=${HOST_2}
docker run \
  -p 2379:2379 \
  -p 2380:2380 \
  --volume=${DATA_DIR}:/etcd-data \
  --name etcd ${REGISTRY}:${ETCD_VERSION} \
  /usr/local/bin/etcd \
  --data-dir=/etcd-data --name ${THIS_NAME} \
  --initial-advertise-peer-urls http://${THIS_IP}:2380 --listen-peer-urls http://0.0.0.0:2380 \
  --advertise-client-urls http://${THIS_IP}:2379 --listen-client-urls http://0.0.0.0:2379 \
  --initial-cluster ${CLUSTER} \
  --initial-cluster-state ${CLUSTER_STATE} --initial-cluster-token ${TOKEN}

# For node 3
THIS_NAME=${NAME_3}
THIS_IP=${HOST_3}
docker run \
  -p 2379:2379 \
  -p 2380:2380 \
  --volume=${DATA_DIR}:/etcd-data \
  --name etcd ${REGISTRY}:${ETCD_VERSION} \
  /usr/local/bin/etcd \
  --data-dir=/etcd-data --name ${THIS_NAME} \
  --initial-advertise-peer-urls http://${THIS_IP}:2380 --listen-peer-urls http://0.0.0.0:2380 \
  --advertise-client-urls http://${THIS_IP}:2379 --listen-client-urls http://0.0.0.0:2379 \
  --initial-cluster ${CLUSTER} \
  --initial-cluster-state ${CLUSTER_STATE} --initial-cluster-token ${TOKEN}
```
 Additionally, I would like to request you to please provide a simpler version of the above commands that can be run to create a 3 node etcd cluster as it is a bit difficult for me to understand these current commands.

**Comment by jmhbnz** [comment]
Hi @Anand11Yash - I believe the instructions you're referring to expect three separate machines, can I clarify if you are running all commands on one machine?

Within linux systems you generally can't have more than one process binding to a port number. When a process binds to a specific port on an IP address, it establishes itself as the receiver for any incoming traffic targeted at that port. If another process attempts to bind to the same port and IP address, it will typically encounter an "Address already in use" error you are seeing.

If you would like to run a local multi member cluster all on one machine you would need to use different ports for each one, alternatively consider following the goreman example in our readme: https://github.com/etcd-io/etcd#running-a-local-etcd-cluster

**Reply by dev-vaayen** [reply]
Thank you for the quick response @jmhbnz. I am running all these commands on one machine by using Docker.

I really appreciate your simple explanation for why the error is taking place in the first place. However, is there any other possible way for making etcd clusters using Docker other than the way it has been explained in the document?

As per my understanding, these instructions are expected to be run on the same machine as each Docker container is to represent a separate instance of the etcd node within the cluster, so that we can simulate a multi-node etcd cluster on a single machine. 

I have specifically been asked to create etcd clusters using Docker containers. Can you please confirm if I will be needed to change the ports for all the nodes when running them in containers?

**Reply by chaochn47** [reply]
Yes. Also remember changing the host to localhost.

**Reply by liangyuanpeng** [reply]
@Anand11Yash  Just change the expose port, like:

```
node1:  `-p 1379:2379 -p 1380:2380 `
node2:  `-p 2379:2379 -p 2380:2380 `
node3:  `-p 3379:2379 -p 3380:2380 `
```
